### PR TITLE
update syn for `remacs-macros`

### PR DESCRIPTION
**Pull request by cantsin**
Partially addresses https://github.com/remacs/remacs/issues/1244

It doesn't look like we can upgrade `syn` for the `remacs-util` crate because newer versions of `syn` use `proc_macro` internally and trying to use these related functions outside of `#[proc_macro_attribute]` or `#[proc_macro_derive(...)]` will give us a `procedural macro API is used outside of a procedural macro` error. (We manually parse `lisp_fn` attributes in `remacs_lib/docfile.rs`)

The older version of `syn` (0.11.11) for the `remacs-util` crate works because it uses a fork of `nom` called [synom](https://crates.io/crates/synom) instead of relying on `proc_macro`; as far as I can see, this fork was abandoned a couple of years ago. 

TL;DR: this PR updates `syn` where-ever possible (only `remacs-macros`); we may need to re-write or re-structure `remacs-util` if we want to update it in the future (perhaps it should utilize `nom` directly).

### CODE REVIEW

**Comment by brotzeit** [comment]
Thanks for the investigation =)

**Comment by cantsin** [comment]
Thanks! What's the etiquette here wrt merging? Can I merge myself after the PR is approved? (Apologies if this was documented somewhere, but I didn't see it in the wiki).

**Comment by brotzeit** [comment]
We let the author do the merge. I usually give others the chance to review my PRs after an approval and wait another few days. But this PR is already 10 days old and I think there won't be another review.
